### ISSUE TITLE
[Feature] Native support in xgrammar for transformers beam search

### ISSUE THREAD

**Issue by SvenPVoigt** [mlc-ai/xgrammar#681]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The latest version does not have this feature.
- [x] 3. I have filled all the required information below.
- [x] 4. The feature request is in English.

### requested feature

A variant of the xgrammar hf.LogitsProcessor that supports beam search in huggingface transformers.

#### Setup XGrammar

Potentially a parameter

```diff
# ... boilerplate code
-xgr_logits_processor = xgr.contrib.hf.LogitsProcessor(compiled_grammar)
+xgr_logits_processor_beams = xgr.contrib.hf.LogitsProcessor(compiled_grammar, stateless=True)
```

**OR** a new class name

```diff
# ... boilerplate code
-xgr_logits_processor = xgr.contrib.hf.LogitsProcessor(compiled_grammar)
+xgr_logits_processor_beams = xgr.contrib.hf.BeamSearchLogitsProcessor(compiled_grammar)
```

#### Huggingface Transformers Generation

```diff
# ... boilerplate code
-model.generate(**inputs, max_new_tokens=32, logits_processor=[xgr_logits_processor]) #num_beams throws error
+model.generate(**inputs, max_new_tokens=32, logits_processor=[xgr_logits_processor_beams], num_beams=15)
```

### Motivation

Constrained generation often lacks accuracy compared to plain feature extraction using vanilla llms because constraints distort the probability distributions. However, with vanilla llms, you don't get alignment across documents.

Beam search improves accuracy by picking the preferred sequence of tokens rather than greedily picking the first best token at each step.

Huggingface already supports beam search, however, it reorders beams. The current xgrammar processor advances one matcher per row and breaks with beam search because beams are reordered. This disconnects the row and corresponding matcher and the fork method can therefore not be used in a logits processor. The only solution at the moment is to rewalk the input tokens on each generation step to re-obtain beam state.

Also, xgrammar has a place for huggingface relevant code in xgrammar.contrib.hf - therefore, I think beam search support could be a natural addition.

### Alternatives

Early work avoids the entire problem of less common tokens by only considering tokenizations of the chosen vocabulary, [De Cao et al., Autoregressive Entity Retrieval, 2021][1].

Recent work develops techniques for normalizing the distributions over a single token [Tran-Thien, Fast, High-Fidelity LLM Decoding, 2024][2].

For my suggestion of beam search, a wrapper can be constructed around the xgrammar logitsprocessor

```python
class WrappedXGrammarProcessor(LogitsProcessor):
    def __init__(self, schema, prompt_len, tokenizer):
        # Compile Grammar

    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor):
        # Blank mask

        # Refill mask - loop over input_ids rows
            # If eos, fill as eos, continue
            # New matcher
            # Rewalk tokens
            # Fill mask

        # Apply mask to scores
        # Return scores
```

Given that there is already code specific to huggingface in xgrammar and this is technically possible with a little rearrangement of where in a loop the matcher lives - is there a place for native huggingface beam search capabilities in xgrammar? I would be happy to implement it if there is interest.

[1]: https://arxiv.org/abs/2010.00904
[2]: https://huggingface.co/blog/vivien/llm-decoding-with-regex-constraints
